### ISSUE TITLE
drishti hunter package 

### ISSUE THREAD

**Issue by headupinclouds** [elucideye/drishti#477]
TODO: drishti/cmake/Config.cmake.in etc

A drishti-assets companion package will be a follow up, but the format needs some thought.

Relates to : https://github.com/elucideye/drishti/issues/301

Revisions to public SDK are expected.

**Comment by ruslo** [comment]
> A drishti-assets companion package will be a follow up, but the format needs some thought

We are creating package from https://github.com/elucideye/drishti-assets ? In this case `drishti_download_data`/`drishti_download_private_data` can be removed (`drishti_download_private_data` isn't used now by the way).

**Comment by headupinclouds** [comment]
> We are creating package from https://github.com/elucideye/drishti-assets

Yes, after drishti hunter packaging, I think that is the right approach -- we can leverage the new Hunter cellar feature.  Thoughts?   (We can create another issue for this, as there are a few open issues related to it.)  For typical use, a compact package containing only the most recent cereal `*.cpb` archives would be used.  A small index file containing key=value pairs (json/xml?) linking model files to specific functions will be needed.  This would be a collection of files necessary to create a `FaceDetectorFactory`.  The Qt resource variant is shown here:

https://github.com/elucideye/drishti/blob/master/src/app/qt/facefilter/QtFaceDetectorFactory.cpp#L34-L49

```
#define DRISHTI_FACE_MEAN_5_POINT "drishti_face_tight_64x64_gray_V5_mean.xml"
#define DRISHTI_FACE_INNER_DETECT "drishti_face_tight_64x64_gray_V5." DRISHTI_ARCHIVE
#define DRISHTI_FACE_INNER "drishti_face_inner." DRISHTI_ARCHIVE
#define DRISHTI_EYE_FULL "drishti_eye_full_npd_eix." DRISHTI_ARCHIVE
```

We would want to archive and exclude older model files in `drishti-assets`.

It might look like this:

face_detector_factory.json:
```
{
    "DRISHTI_FACE_INNER_DETECT": "drishti_face_tight_64x64_gray_V5.cpb"
    "DRISHTI_EYE_FULL": drishti_eye_full_npd_eix.cpb"
    "DRISHTI_FACE_INNER": "drishti_face_inner.cpb"
    "FACE_MEAN_5_POINT": "drishti_face_tight_64x64_gray_V5_mean.xml"
}
```

These files could optionally be bundled (perhaps via CMake tar) to reduce the # of command line options and configuration that is required to use this in various console apps and unit tests.  Eventually,`cpp_rsc` could be used to provide built in resources (similar to Qt) if needed.

**Comment by ruslo** [comment]
> For typical use, a compact package containing only the most recent cereal *.cpb archives would be used

So it's not just "unpack archive", we have to process unpacked files on-the-fly and introduce extra files?

> The Qt resource variant is shown here

Quite tricky to generalize. Need to think.

**Comment by ruslo** [comment]
In this code:
* https://github.com/elucideye/drishti/blob/d1cd108be996fd7c39dc6bd31307525ff1fcb059/cmake/Modules/drishti_split_debug_symbols.cmake#L45-L50

we are installing `.build-id/ab/cdef1234.debug`, we should add `drishti` somewhere, like `.build-id/ab/drishti-cdef1234.debug`.

**Comment by headupinclouds** [comment]
note #482

**Comment by ruslo** [comment]
Summarizing issues I currently have while building with Hunter default config.

(Note: I'm only **building** with GCC on Linux. There may be more issues for **other platforms** or **runtime** errors).

### OpenCV

Error when using default OpenCV 3.2.0-p2 instead of custom OpenCV 3.0.0-p11:
```
/.../drishti/src/lib/drishti/rcpr/cprApply.cpp:
    In member function ‘int drishti::rcpr::CPR::cprApplyTree(const cv::Mat&, const drishti::rcpr::CPR::RegModel&, const Vector1d&, drishti::rcpr::CPR::CPRResult&, bool) const’:

/.../drishti/src/lib/drishti/rcpr/cprApply.cpp:221:70:
    error: call of overloaded ‘cprApplyTree(<brace-enclosed initializer list>, const drishti::rcpr::CPR::RegModel&, const Vector1d&, drishti::rcpr::CPR::CPRResult&, bool&)’ is ambiguous
     return cprApplyTree({ I, mask }, regModel, pIn, result, doPreview);
                                                                      ^
/.../drishti/src/lib/drishti/rcpr/cprApply.cpp:218:5:
    note: candidate: int drishti::rcpr::CPR::cprApplyTree(const cv::Mat&, const drishti::rcpr::CPR::RegModel&, const Vector1d&, drishti::rcpr::CPR::CPRResult&, bool) const
 int CPR::cprApplyTree(const cv::Mat& I, const RegModel& regModel, const Vector1d& pIn, CPRResult& result, bool doPreview) const
     ^

In file included from /.../drishti/src/lib/drishti/rcpr/cprApply.cpp:12:0:
/.../drishti/src/lib/drishti/rcpr/CPR.h:304:9: note: candidate: int drishti::rcpr::CPR::cprApplyTree(const drishti::rcpr::ImageMaskPair&, const drishti::rcpr::CPR::RegModel&, const Vector1d&, drishti::rcpr::CPR::CPRResult&, bool) const
     int cprApplyTree(const ImageMaskPair& Is, const RegModel& regModel, const Vector1d& p, CPRResult& result, bool preview = false) const;
         ^
```

### spdlog

Error when using default spdlog 0.13.0-p0 instead of custom spdlog 1.0.0-p0:
```
/.../drishti/src/lib/drishti/acf/GPUACF.cpp: In lambda function:
/.../drishti/src/lib/drishti/acf/GPUACF.cpp:788:34: error: no matching function for call to ‘spdlog::logger::info()’
             impl->m_logger->info() << "TIMING:" << tag << ":" << ss.str() << ";total=" << elapsed;
                                  ^
```

Seems that stream-like `info() << "TIMING"` usage is no longer supported:
* https://github.com/gabime/spdlog#usage-example

Note that [1.0.0-p0](https://github.com/hunter-packages/spdlog/releases/tag/v1.0.0-p0) was created by me. There was no official versions released before so I've mistakenly started with `1.0`.

First official version:
* https://github.com/gabime/spdlog/releases/tag/v0.9.0

Current official version:
* https://github.com/gabime/spdlog/releases/tag/v0.13.0

It means that `0.13.0 > 1.0.0` in fact here.

### xgboost

Currently version `0.40-p5` from branch `hunter` used:
* https://github.com/hunter-packages/xgboost/commit/05fc548e53b63bdd2d80c8fb3e9a6d0e428424d1

Common commit with upstream https://github.com/dmlc/xgboost is:
* https://github.com/dmlc/xgboost/commit/bda3282f6d1e548fb9081898da4058cdd8abc8d2 (which is `v0.60+`)

Compare:
* https://github.com/hunter-packages/xgboost/compare/bda3282f6d1e548fb9081898da4058cdd8abc8d2...05fc548e53b63bdd2d80c8fb3e9a6d0e428424d1

So the main question here is can we make this version default? It means if user of upstream `xgboost` will use Hunter version then will he experience build errors or non-standard behavior? If it's just a kind of extention then I guess it's okay to make it default.

**Comment by headupinclouds** [comment]
### OpencV

> Error when using default OpenCV 3.2.0-p2 instead of custom OpenCV 3.0.0-p11

I don't understand this one.  The `{ I, mask }` initialization probably needs to be replaced w/ an explicit type `ImageMaskPair(I,mask)` due to some OpenCV changes.

### spdlog

> Seems that stream-like info() << "TIMING" usage is no longer supported

We can update to the latest syntax.  For reference (from their webpage) this will entail changing

OLD:
```
logger->info() << "TIMING";
```
NEW:
```
logger->info("TIMING")
```

I'm curious what issues the stream style operators introduced.

###  xgboost

> So the main question here is can we make this version default

TL;DR: We will need to keep 0.40-p5 w/ current CMAKE_ARGS for now. 

Drishti needs the modified xgboost.  Maybe it is better to rename this to indicate fork specific modifications.  Something like xgboost-lite.  As you have mentioned, xgboost-lite and xgboost would not be compatible for use in the same project.  It would be nice to have a standard xgboost build in hunter for people to use as well (minimal CMake hunterizations).  I would like to update the xgboost version in hunter-packages, but for drishti, we need to keep the extensions, primarily cereal portable binary archives w/ half precision floating point serialization (we can drop boost support).  After applying the patches to the newest xgboost, we will need to retrain some models (using cereal archives) based on xgboost data structure updates.  I can create some issues around this.  Unfortunately, this one is a non-trivial change.  I've created a small standalone integration library to encapsulate this layer here: https://github.com/elucideye/xgboostcpp.  We can work through changes there and integrate that in drishti when it is ready.  This version has a pointer to the same xboost drishti is using, but with cleaner separation of cereal and boost serialization.  Removing boost stuff will make the serialization much cleaner.  (Trying to maintain cereal and boost side-by-side is a little messy.)

**Comment by ruslo** [comment]
> Maybe it is better to rename this to indicate fork specific modifications

I think it's a bad option. If somebody want to use both packages - one package will overwrite another leading to invalid `<root>` state (they will write same files to the same place like `<root>/include/xgboost/*.h`). And it may happens implicitly when `A` depends on `xgboost`, `B` depends on `xgboost-lite`. User want to have `A` and `B`... Currently Hunter doesn't control such conflicts at all. When this control will be added, it will be used for the packages that implement **same** API differently. Meaning packages use stable API but user can for any reason change the exact implementation. An example of such API is C++ standard: we can use Clang + libcxx or GCC + libstdcxx. API should be stable and even if implementations differ the final visible behavior should be the same, so it's still not the case for `xgboost` vs `xgboost-lite`.

I see next options here:

1. Customize CMake option in such way that default configuration confirms default `xgboost`. I.e. if no custom `CMAKE_ARGS` present, behavior and API for `xgboost` will be the same as for upstream. But for the Drishti users we have to explicitly state that you have to create local `config.cmake` and add something like 
```cmake
hunter_config(xgboost VERSION ${HUNTER_xgboost_VERSION} CMAKE_ARGS XGBOOST_USE_BOOST=YES)
```
However it still means user can't have both `xgboost` and patched `xgboost` in one project, but at least it's clear from configuration now.

2. Create C++ wrapper for custom features.

I haven't dived into details but as an example in this code:
```cpp
class GBLinear : public IGradBooster {
   virtual void LoadModel(utils::IStream &fi, bool with_pbuffer) {
#if XGBOOST_DO_LEAN
     assert(false);
#else
     model.LoadModel(fi);
     model.LoadModel(fi);
#endif
   }
}
```

Can we do:
```cpp
class GBLinearCustom: public GBLinear {
   virtual void LoadModel(utils::IStream &fi, bool with_pbuffer) {
     assert(false);
   }
}
```

And let Drishti use `GBLinearCustom` instead of `GBLinear`. Is it an option?


As a general note local config for Hunter and custom versions works well if the final goal is to build executable. If you want to build library that will be reused in different projects along with other libraries then custom configs doesn't scale.

**Comment by headupinclouds** [comment]
> What is the reason of adding XGBOOST_DO_LEAN? Size of final binary? 

Yes.

> Will we achieve the same using LTO build?

From the type of changes I recall making, I think `LTO` w/ `-ffunction-sections` and `-fdata-sections` along with `MinSizeRel` should be fairly close to `XGBOOST_DO_LEAN=ON`.  Much of those preprocessor blocks are about stripping out the training functions, since only prediction is required to run the trained models.   We can test it.

**Comment by headupinclouds** [comment]
> However all values hardcoded in C++ code in file src/utils/io.h:

That's my fault.  I've cleaned up the xgboost stuff in the submodule pointer for `xgboostcpp` here:

https://github.com/hunter-packages/xgboost/commits/eb4af799dbcac4166a6241abfcf4e8b55bc1df6c

I believe all the macro/option stuff is consistent in that package.  It also allows one to toggle `cereal` and `boost` serialization: `XGBOOST_USE_BOOST` and `XGBOOST_USE_CEREAL`.  It will be helpful to move to this interface layer at some point in the future.

This module will break some of the trained models, so integration is a slightly larger effort.

**Comment by headupinclouds** [comment]
> XGBOOST_USE_BOOST extends API. It means we can add this feature without breaking upstream.

Ok.  We should make sure that a boost package is not created via `hunter` if `XGBOOST_USE_BOOST=OFF`.

**Comment by ruslo** [comment]
> > What is the reason of adding XGBOOST_DO_LEAN? Size of final binary?

> Yes.

Okay, it seems for me that we are close to resolving the issue :) My main thought here now is that we can submit `XGBOOST_USE_BOOST` without breaking standard API, use `XGBOOST_DO_LEAN=OFF`/`XGBOOST_USE_HALF=OFF` for default and add optimization notes in README for advanced users about `XGBOOST_DO_LEAN=ON`/`XGBOOST_USE_HALF=ON`. This makes Drishti works from-the-box, optimal for advanced users and we don't break xgboost API.

What do you think?

The only place I'm worrying now is this code:
```cpp
#if XGBOOST_DO_LEAN
      // initialize model
      this->InitObjGBM();
      // reset the base score
      mparam.base_score = obj_->ProbToMargin(mparam.base_score);
      // initialize GBM model
      gbm_->InitModel();
#else
     this->InitTrainer(); // <<<<<<<<<<< not called if XGBOOST_DO_LEAN defined
     // initialize model
     this->InitObjGBM();
     // reset the base score
     mparam.base_score = obj_->ProbToMargin(mparam.base_score);
     // initialize GBM model
     gbm_->InitModel();
#endif
```
Is `this->InitTrainer()` call removed for the purpose? What if we put it back?

**Comment by headupinclouds** [comment]
> XGBOOST_SQUEEZE used only if XGBOOST_USE_BOOST defined so it will not break upstream API too. Also is this an optimization? If user will not use this macro the behavior of Drishti still be correct?

This will enable the half precision floating point serialization, which is really a sub-option for the `XGBOOST_USE_BOOST` option (i.e., there is no half precision serialization support for the default xgboost serialization routines).  All of the trained drishti models currently use the `boost` and `cereal` serialization (where syntax between libs is almost compatible, but not entirely) as well as the the half precision floating point storage.  This means that both`XGBOOST_USE_BOOST=ON` and `XGBOOST_SQUEEZE=ON` are required to load the models.  The original xgboost code and integration layer in drishti should be functional without these changes if one trained their own model, although the code would need to be tested to confirm this (it should work).  In the current `hunter` version `XGBOOST_USE_BOOST` is required to enable the cereal serialization.  This is a pain, since it will unnecessarily add boost if it isn't needed.  The version here https://github.com/hunter-packages/xgboost/commits/eb4af799dbcac4166a6241abfcf4e8b55bc1df6c fixes that.

**Comment by headupinclouds** [comment]
> My main thought here now is that we can submit XGBOOST_USE_BOOST without breaking standard API, use XGBOOST_DO_LEAN=OFF/XGBOOST_USE_HALF=OFF for default and add optimization notes in README for advanced users about XGBOOST_DO_LEAN=ON/XGBOOST_USE_HALF=ON. This makes Drishti works from-the-box, optimal for advanced users and we don't break xgboost API.

Kind of.  It would create a *valid* out-of-the-box build, but it wouldn't be able to load the current models that are provided for unit tests and applications.  In practice, I think most people will want pre-trained models.  I guess it would just be a matter of documenting LOCAL config.cmake requirements for these flags.

**Comment by headupinclouds** [comment]
> Is this->InitTrainer() call removed for the purpose? What if we put it back?

I need to check, but the training code is all removed in `XGBOOST_DO_LEAN=ON`.

**Comment by ruslo** [comment]
> Much of those preprocessor blocks are about stripping out the training functions, since only prediction is required to run the trained models.

Sounds like something useful that can be contributed upstream, e.g. XGBOOST_PREDICTION_ONLY.

> All of the trained drishti models currently use the boost and cereal serialization (where syntax between libs is almost compatible, but not entirely) as well as the the half precision floating point storage. This means that both XGBOOST_USE_BOOST=ON and XGBOOST_SQUEEZE=ON are required to load the models.

We are getting models from https://github.com/elucideye/drishti-assets repository currently, right? From Hunter point of view solution is to add model trainer package `drishti_model_trainer` which will generate and install these models. Since CMAKE_ARGS are global `hunter_add_package(xgboost)` will use the same parameters for both generation and usage. For iOS/Android we will need https://github.com/ruslo/hunter/issues/495 feature. It may be not very convenient in practice if training take a long time, if there is no package in binary cache user have to run training from scratch.

**Comment by headupinclouds** [comment]
> From Hunter point of view solution is to add model trainer package drishti_model_trainer which will generate and install these models

In general, I think this may be difficult, but it is a good longer term goal.  This project 

https://github.com/wayaai/SimGAN

could be useful to provide a path forward.  It is an implementation of an Apple paper that uses adversarial networks to synthesize realistic eye images.

I've updated #479 .  

This project uses  tensorflow.  See https://github.com/ruslo/hunter/issues/667

**Comment by headupinclouds** [comment]
#512

**Comment by ruslo** [comment]
Added to Hunter:
* https://docs.hunter.sh/en/latest/packages/pkg/drishti.html

Doing upload at this moment:
* https://travis-ci.org/ingenue/hunter/builds/290845085
Should not take a lot of time. The main issue was about iOS toolchain and it looks good now.

Example should be tweaked for Android and Windows:
* dlib version (waiting for Android NDK 17): https://github.com/ruslo/hunter/commit/08d25c51e8fa3f3fcfa73f655fe3a7d85d1b4109
* OpenCV conversion `double` to `float` error (fix Drishti?): https://github.com/ruslo/hunter/commit/273523aec9fa2d94c655a7953eec2be79c59938a

**Comment by headupinclouds** [comment]
> OpenCV conversion `double` to `float` error (fix Drishti?): https://github.com/ruslo/hunter/commit/273523aec9fa2d94c655a7953eec2be79c59938a

This one isn't technically an OpenCV issue, but is related to a default narrowing-warning-as-error in MSVC builds of drishti.  I pushed a fix here: https://github.com/elucideye/drishti/pull/596 .  I didn't look into why changing the OpenCV version fixes this, but I'm guessing it is related to some subtle changes in the constructor.

**Comment by headupinclouds** [comment]
> Added to Hunter:

Thanks.  It is nice to see the first end-to-end in place.

**Comment by ruslo** [comment]
> Goal: https://github.com/elucideye/drishti_hunter_test

Reopen.

**Comment by headupinclouds** [comment]
quit updates + cleanup: https://github.com/elucideye/drishti_hunter_test/pull/3

**Comment by ruslo** [comment]
Tests passed: https://github.com/elucideye/drishti_hunter_test/pull/9
